Allow null values in XWPFParagraph.get/setAlignment().

jaragunde opened the pull request:
A null value in this field would indicate that the paragraph should follow the alignment provided by the style hierarchy.

Fixes: https://bz.apache.org/bugzilla/show_bug.cgi?id=69720

jaragunde committed:
Allow null values in XWPFParagraph.get/setAlignment().

A null value in this field would indicate that the paragraph should
follow the alignment provided by the style hierarchy.

Fixes: https://bz.apache.org/bugzilla/show_bug.cgi?id=69720

jaragunde commented:
Hi! A gentle ping. Anything I can do on my side to make review easier? Maybe I missed some contribution step (e.g. signing a contributor agreement)? Thanks in advance for your time 🙂

pjfanning commented on poi-ooxml/src/main/java/org/apache/poi/xwpf/usermodel/XWPFParagraph.java:
```diff
@@ -506,13 +506,18 @@ public ParagraphAlignment getAlignment() {
-        STJc.Enum en = STJc.Enum.forInt(align.getValue());
-        jc.setVal(en);
+        if (align == null) {
+            pr.unsetJc();
```
* is there some way that getCTPPr(init = false) could be used here? we have equivalent code elsewhere
* we don't want to create a CTPPr instance if none already exists for the null input case

jaragunde replied:
I added `getCTPPr(create)` in one of the patches I just uploaded, and I kept `getCTPPr()` with no arguments to call `getCTPPr(true)`, preserving the original behavior. I looked for similar implementations and found `XSLFPlaceholderDetails.getCTPlaceholder(create)`, although this kind of behavior didn't seem to be very widely used in the codebase.

I can submit it as a separate patch/PR, whatever is most convenient for you. Thanks!

pjfanning commented on poi-ooxml/src/main/java/org/apache/poi/xwpf/usermodel/XWPFParagraph.java:
```diff
@@ -488,11 +488,11 @@ public String getFootnoteText() {
      * paragraph.
      *
-     * @return the paragraph alignment of this paragraph.
+     * @return the paragraph alignment of this paragraph. Can be null if not set.
```
This is a breaking change. I know you don't like the way things work but unfortunately, in a 20 year old project with lots of users, we can't just change API behaviours. We have no major releases planned, just patch releases - so if you want to break the API, this change will need to wait.

Can you instead consider adding a new method `public boolean isAlignmentSet()`? New methods won't cause backward compatibility issues.

jaragunde replied:
Thanks a lot for the insight! You are totally right, backwards API compatibility must be considered, and a method that starts returning null can cause a lot of headaches... I will implement this behavior as a new method instead.

What do you think about allowing `setAlignment(null)`? It sounds less dangerous from the compatibility POV, but maybe it looks a bit weird? Better with a new method `unsetAlignment()`?

pjfanning replied:
I think your change for `setAlignment(null)` is actually ok. That change is more of a bug fix because `STJc.Enum en = STJc.Enum.forInt(align.getValue())` (existing code) will NPE if align=null.

jaragunde replied:
I reverted `getAlignment()` to the original behavior and implemented `isAlignmentSet()` as discussed. Thanks!

jaragunde committed:
Revert getAlignment() and implement isAlignmentSet() instead.

jaragunde committed:
Replace test case file.

jaragunde committed:
Implement XWPFParagraph.getTCPPr(create).

It lets the caller choose if a new PPr should be created in case it
doesn't exist.

pjfanning committed:
use getCTPPr(boolean)

pjfanning committed:
compile issue

pjfanning reviewed (approved):
lgtm - thanks

jaragunde commented:
Thanks for your review!
